### PR TITLE
Initial implementation

### PR DESCRIPTION
**Pull request by kopiczko**
Towards https://github.com/giantswarm/giantswarm/issues/20946

There is much to improve here, but I wanted to keep to implementation as simple as possible for the first iteration.

Output for config:

```yaml
- chart: "cilium"
  versions: ">= 1.11.2"
  srcHelmRepo: "https://helm.cilium.io"
  dstCatalog: "default-test"
- chart: "does-not-exist"
  versions: ">= 1.2.3"
  srcHelmRepo: "https://example.com"
  dstCatalog: "default-test"
```

```
====> Cloning destination catalogs
--> Shell: git clone --depth=1 "git@github.com:giantswarm/default-test-catalog.git" ./tmp-recharter-catalog
Cloning into './tmp-recharter-catalog'...
remote: Enumerating objects: 1483, done.
remote: Counting objects: 100% (1483/1483), done.
remote: Compressing objects: 100% (1335/1335), done.
remote: Total 1483 (delta 236), reused 1371 (delta 145), pack-reused 0
Receiving objects: 100% (1483/1483), 23.36 MiB | 1.02 MiB/s, done.
Resolving deltas: 100% (236/236), done.
====> Synchronizing "cilium" with versions ">= 1.11.2" from https://helm.cilium.io to "default-test-catalog"
--> Shell: mkdir ./tmp-recharter-tarballs
--> Chart cilium@1.11.2 found in the "default-test-catalog" catalog, skipping
--> No new releases downloaded from https://helm.cilium.io
====> Synchronizing "does-not-exist" with versions ">= 1.2.3" from https://example.com to "default-test-catalog"
error: parsing source (https://example.com/index.yaml) index: unmarshaling index YAML: error converting YAML to JSON: yaml: line 11: mapping values are not allowed in this context
```

### CODE REVIEW

File: pkg/shell/shell.go [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+	for _, o := range b.options {
+		o.PreRun(cmd)
+	}
+	err := cmd.Run()
+	var eerr *exec.ExitError
+	if err != nil && errors.As(err, &eerr) {
+		os.Exit(eerr.ExitCode())
+	} else if err != nil {
+		fmt.Fprintf(os.Stderr, "error executing command: %s\n", err)
+		os.Exit(1)
```
Comment (kopiczko) [inline pkg/shell/shell.go:52]: I understand this `os.Exit` call is a bit inconsistent with the rest of error handling, but it results in huge code read-ability improvement.

File: LICENSE [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+      To apply the Apache License to your work, attach the following
+      boilerplate notice, with the fields enclosed by brackets "{}"
+      replaced with your own identifying information. (Don't include
+      the brackets!)  The text should be enclosed in the appropriate
+      comment syntax for the file format. We also recommend that a
+      file or class name and description of purpose be included on the
+      same "printed page" as the copyright notice for easier
+      identification within third-party archives.
+
+   Copyright 2016 - 2021 Giant Swarm GmbH
```
Comment (pipo02mix) [inline LICENSE:189]:
```suggestion
   Copyright 2016 - 2022 Giant Swarm GmbH
```

File: charts.yaml [hunk 3]
```diff
@@ -0,0 +1,4 @@
+- chart: "cilium"
+  versions: ">= 1.11.2"
+  srcHelmRepo: "https://helm.cilium.io"
+  dstCatalog: "default-test"
```
Comment (pipo02mix) [inline charts.yaml:4]: is it much harder to use URL format like `srcHelmRepo` so it is more generic?
```suggestion
  dstHelmRepo: "https://github.com/giantswarm/default-test"
```

File: cmd/run/run.go [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+	// Update the destination catalog index.yaml and move the downloaded tarballs.
+
+	shell.Cmdf("helm repo index --url=%q --merge=./tmp-recharter-catalog/index.yaml ./tmp-recharter-tarballs", catalogURL).OrDie()
+	shell.Cmdf("cp -a ./tmp-recharter-tarballs/* ./tmp-recharter-catalog").OrDie()
+
+	// Commit and push the destination catalog.
+
+	shell.Cmdf("git -C ./tmp-recharter-catalog add -A").OrDie()
+	shell.Cmdf("git -C ./tmp-recharter-catalog commit -m %q", "Pull "+config.SrcHelmRepo+" releases").OrDie()
+	shell.Cmdf("git -C ./tmp-recharter-catalog push").OrDie()
```
Comment (pipo02mix) [inline cmd/run/run.go:142]: Dont we use a PR?

File: pkg/shell/shell.go [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+
+	cmd := exec.Command("sh", "-c", b.command)
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+
+	for _, o := range b.options {
+		o.PreRun(cmd)
+	}
+	err := cmd.Run()
+	var eerr *exec.ExitError
```
Comment (tfussell) [inline pkg/shell/shell.go:47]: Can we call this exitErr or execErr?

**Review by tfussell — APPROVED** [review]
I don't like the use of `sh -c` and relying on executables existing and this doesn't follow some of our `fmt` guidelines, but it's fine for a first step.

**Comment by kopiczko** [comment]
I won't continue with that. It's better to use kustomization to generate helm charts IMO.

**Pull request closed without merging**